pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,272 @@
+	KeyFile string
+}
+
+func init() {
+	_ = corev1.AddToScheme(runtimeScheme)
+	_ = admissionregistrationv1beta1.AddToScheme(runtimeScheme)
+	// defaulting with webhooks:
+	// https://github.com/kubernetes/kubernetes/issues/57982
+	_ = v1.AddToScheme(runtimeScheme)
+}
+
+// NewWebhook creates a new instance of a webhook.
+func NewWebhook(p WParameters, kubeClient kubernetes.Interface, openebsClient clientset.Interface) (*Webhook, error) {
+
+	pair, err := tls.LoadX509KeyPair(p.CertFile, p.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+	wh := &Webhook{
+		Server: &http.Server{
+			Addr:      fmt.Sprintf(":%v", p.Port),
+			TLSConfig: &tls.Config{Certificates: []tls.Certificate{pair}},
+		},
+		KubeClient: kubeClient,
+		Clientset:  openebsClient,
+	}
+	return wh, nil
+}
+
+func admissionRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	// skip special kubernetes system namespaces
+	for _, namespace := range ignoredList {
+		if metadata.Namespace == namespace {
+			glog.Infof("Skip validation for %v for it's in special namespace:%v", metadata.Name, metadata.Namespace)
+			return false
+		}
+	}
+	return true
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, metadata)
+	glog.V(4).Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
+	return required
+}
+
+// validate validates the persistentvolumeclaim(PVC) delete request
+func (wh *Webhook) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	req := ar.Request
+	response := &v1beta1.AdmissionResponse{}
+
+	// validates only if requested operation is DELETE and request kind is PVC
+	if ar.Request.Operation != v1beta1.Delete && req.Kind.Kind != "PersistentVolumeClaim" {
+		response.Allowed = true
+		return response
+	}
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	// TODO* use raw object once validation webhooks support DELETE request
+	// object as non nil value https://github.com/kubernetes/kubernetes/issues/66536
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := wh.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("error retrieving PVC: %v", err.Error()),
+		}
+		return response
+	}
+
+	if !validationRequired(ignoredNamespaces, &pvc.ObjectMeta) {
```

Comment:
do we need to have `validationRequired` kind of thing only for cstor volumes?
I mean - jiva volumes doesn't need to go through this.. right?

Reply:
Ideally we should not provision something in kube system related namespaces.  validation will skip as part of above validationrequired.